torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1338,6 +1338,13 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
     OpType opType,
     const char* profilingTitle) {
 
+  // A hack to copy lazy to cuda to bypass following checks.
+  if (inputs.front().device().type() == at::kLazy) {
```

Comment:
so is this the part you mentioned about falling back before the 'real' part of DDP happens?

you first do all of the compute on lazy device, and then here you materialize all the tensors and run eager DDP?

'inputs' here refers to all the grad tensors that ddp wants to bcast/allreduce?

Reply:
Yes! Even though I think ideally we need to also move the grads back to lazy for the optimizer, yet I don't get any complains within the optimizer. (I'm double checking.)